syntax = "proto3";

package protos;

message Message {
  Type type = 1;
  optional UserData user_data = 2;
  optional ErrorReply error_message = 4;
  optional AlbumName album_name = 5;
  optional Albums albums = 7;
  optional Album album = 8;
  optional NodeInfo node_info = 9;
  repeated NodeInfo nodes_infos = 10;
  optional uint64 token = 11;
  repeated NodeIp nodes_ips = 12;
  optional NewClient new_client = 13;
  optional bool is_last = 6;
  optional LeaveData leave_data = 3;
}

// C2S -> Client to Server
// S2C -> Server to Client

// Type -> IN Data | Out Data
enum Type {
  SUCESIUM = 3; // S2C -> None
  ERRORREPLY = 4; // S2C -> error_message
  ALBUMS = 8; // S2C -> albums
  REGISTER = 0; // C2S -> user_data | {SUCESIUM | ERRORREPLY}
  LOGIN = 1; // C2S -> user_data | {SUCESIUM | ERRORREPLY}
  LOGOUT = 2; // C2S -> None | {SUCESIUM | ERRORREPLY}
  ALBUMSLIST = 5; // C2S -> None | ALBUMS
  ALBUMCREATE = 6; // C2S -> album_name | {SUCESIUM | ERRORREPLY}
  ALBUM = 7; // S2C -> album
  ALBUMEDIT = 9; // C2S -> album_name | {ALBUM | ERRORREPLY | NEWCLIENT}
  STARTENTRANCE = 10;  //C2S -> node_info | {ERRORREPLY | nodes}
  ENDENTRANCE = 11; //C2S -> node_info | {SUCESIUM | ERRORREPLY}
  NODESINFO = 12; // S2C -> nodes_infos
  READ = 13; // C2S -> token | { ERRORREPLY | NODEIP }
  WRITE = 14; // C2S -> token | { ERRORREPLY | NODEIP }
  NODEIP = 15; // S2C -> nodes_ips
  NEWCLIENT = 16; // S2C -> new_client
  ISLASTREPLY = 17; // S2C -> is_last
  ISLAST = 18; // C2S -> album_name | {ERRORREPLY | ISLASTREPLY}
  LEAVE = 19; // C2S -> leave_data | {ERRORREPLY | SUCESIUM}
}

message UserData {
  string username = 1;
  string password = 2;
}

message ErrorReply {
  string message = 1;
}

message AlbumName {
  string name = 1;
}

message Albums {
  repeated string names = 1;
}


message Classifications {
  string username = 1;
  uint32 value = 2;
}

message File {
  string name = 1;
  uint64 hash = 2;
  repeated Classifications classifications = 3;  
}

message Client {
  string username = 1;
  string ip = 2;
  uint32 port = 3;
}

message NewClient {
  repeated Client clients = 1;
  uint32 clock = 2;
}

message LeaveData {
  string name = 1;
  uint32 clock = 2; 
}

message Album {
  repeated File files = 1;
  repeated string users = 2; 
}


message NodeInfo{
  string ip = 1;
  uint32 port = 2;
  repeated uint64 tokens = 3;
}

message NodeIp{
  string ip = 1;
  uint32 port = 2;
}